ludwig/trainers/trainer.py

```diff
@@ -272,6 +272,10 @@ def closure():
                     targets, model_outputs, self.regularization_type, self.regularization_lambda
                 )
                 loss = loss / self.gradient_accumulation_steps
+                try:
+                    loss.requires_grad = True
+                except RuntimeError:
+                    pass
```

Comment:
Why do we have to add this?

Reply:
This was a workaround I had because, when the lora weights were loaded in, some of the loss functions did not have requires_grad set to True. However, without the try-except block, this would try to set requires_grad to True for some intermediate loss functions, which isn't valid.

Reply:
Questions:
1. Why would loading lora weights have loss functions with `requires_grad != True`?
2. What does the training error look like when some loss functions have `requires_grad != True`?
3. Do you have an example of an intermediate loss function that raises an error when you try to set `requires_grad=True`?

At least, we should be adding a comment explaining why this is here, e.g.

> "When loading adpater weights from huggingface or a local path, some of the loss functions do not have `requires_grad=True`. `requires_grad=True` is necessary for back-propogation, but we wrap this in a try/except because some intermediate losses like __ raise an error if `requires_grad` is explicitly set to `True` in this way."

Reply:
Can you elaborate a bit more? What are the intermediate loss functions? I'm also not totally sure how wrapping the model with the adapter is causing these issues?

Reply:
I agree with @justinxzhao and have the same questions as well

Reply:
1. I'm not quite sure on why this was the case. My hypothesis is that, when lora weights are loaded, they overwrite some of the parameters of certain layers.
2. When some loss functions have requires_grad != True, training stops and errors out.
3. I don't have an example of this, but I think my terminology was incorrect here. Specifically, requires_grad can only be changed on leaf variables, so if there is a loss function that is not a leaf node, setting requires_grad=True would cause an error.

Reply:
By default, when a PEFT pretrained adapter is loaded in, is it set to training mode or inference mode? Does that maybe have something to do with requires_grad not being set?

Reply:
Good call. Some useful references:
- https://discuss.pytorch.org/t/check-if-model-is-eval-or-train/9395
- https://stackoverflow.com/questions/60018578/what-does-model-eval-do-in-pytorch